### PR TITLE
Perform handshake on first read/write

### PR DESCRIPTION
**Pull request by kevmo314**
#### Description
Updates the connection to perform a handshake on first read/write instead of on accept. 
#### Reference issue
https://github.com/pion/dtls/issues/279

### CODE REVIEW

**Comment by Sean-Der** [comment]
PR looks beautiful to me! Matches [Upstream](https://pkg.go.dev/crypto/tls#Conn.Handshake) and fixes biggest issue with lib at the moment :)

File: config.go [hunk 1]
```diff
@@ -118,13 +118,10 @@ type Config struct {
...
-	// ConnectContextMaker is a function to make a context used in Dial(),
-	// Client(), Server(), and Accept(). If nil, the default ConnectContextMaker
-	// is used. It can be implemented as following.
+	// ConnectContextMaker is no longer used. It is kept for compatibility
+	// reasons and will be removed in a future release.
 	//
-	// 	func ConnectContextMaker() (context.Context, func()) {
-	// 		return context.WithTimeout(context.Background(), 30*time.Second)
-	// 	}
+	// Deprecated: Use the context parameter in [HandshakeContext] instead.
```
Comment (Sean-Der) [inline config.go:124]: I think we should kill this now. We are going with a major release, lets break now :)

File: conn.go [hunk 2]
```diff
@@ -336,17 +331,35 @@ func ServerWithContext(ctx context.Context, conn net.PacketConn, rAddr net.Addr,
...
 	}
-	dconn, err := createConn(conn, rAddr, config, false)
-	if err != nil {
-		return nil, err
-	}
-	return handshakeConn(ctx, dconn, config, false, nil)
+	return createConn(conn, rAddr, config, false, nil)
+}
+
+// DialWithContext connects to the given network address and establishes a DTLS
```
Comment (Sean-Der) [inline conn.go:337]: I am in support of getting rid of all of these.

File: examples/listen/cid/main.go [hunk 3]
```diff
@@ -32,13 +26,9 @@ func main() {
 			fmt.Printf("Client's hint: %s \n", hint)
 			return []byte{0xAB, 0xC1, 0x23}, nil
 		},
-		PSKIdentityHint:      []byte("Pion DTLS Server"),
-		CipherSuites:         []dtls.CipherSuiteID{dtls.TLS_PSK_WITH_AES_128_CCM_8},
-		ExtendedMasterSecret: dtls.RequireExtendedMasterSecret,
-		// Create timeout context for accepted connection.
-		ConnectContextMaker: func() (context.Context, func()) {
```
Comment (Sean-Der) [inline examples/listen/cid/main.go:39]: Is it easy/possible to do 30 second timeout still with new APIs?

**Comment by Sean-Der** [comment]
Damn @kevmo314 really exceptional PR, I am so grateful for your help.

If you are in support I say we delete/break as much as we can now. Since we are doing a major release try to go with zero baggage. 

Also would be nice to have zero functionally change in examples. Users can reference them when updating their code.

File: examples/listen/cid/main.go (hunk 3 above)
Comment (kevmo314) [reply examples/listen/cid/main.go:39]: Yes, you would essentially call `conn.HandshakeContext(<30 second timeout context>)` manually before the first read/write. In other words, the sequence of calls would be something like:

```go
for {
  conn := lis.Accept()
  go func() {
    if err := conn.Handshake(<30 second timeout context>); err != nil {
      ...
    }
    // regular read/write stuff
  }()
}
```

**Comment by kevmo314** [comment]
> Also would be nice to have zero functionally change in examples. Users can reference them when updating their code.

Sounds good, I'll update them to include the timeout context.

**Pull request merged**